src/base/bittorrent/tracker.h

```diff
@@ -49,21 +50,76 @@ namespace Http
 
 namespace BitTorrent
 {
-    struct Peer
+    class Peer
     {
-        QString ip;
-        QString peerId;
-        int port;
+    public:
+        Peer() = default;
+        Peer(QString&& ip, QString&& announce_ip, std::string&& peer_id, int port, QByteArray&& uid);
+
+        Peer(const Peer&) = default;
+        Peer(Peer&&) = default;
+        Peer& operator=(const Peer&) = default;
+        Peer& operator=(Peer&&) = default;
+
+        const QString& getIp() const;
 
         bool operator!=(const Peer &other) const;
         bool operator==(const Peer &other) const;
-        QString uid() const;
-        libtorrent::entry toEntry(bool noPeerId) const;
+        const QByteArray& uid() const;
+        // converts to appropriate entry taking into account whether 'clientIp' is a public ip address,  i.e. provides
+        // public ip if 'clientIP' belongs to a public network, otherwise gives connection ip
+        libtorrent::entry toEntryForClient(bool noPeerId, const QString& clientIp) const;
```

Comment:
And then it all makes the PeerBuilder class superfluous!
If you apply my suggestions, you will significantly reduce the amount of code.